### PR TITLE
Avoid shadowing of variables [blocks: #2310]

### PR DESCRIPTION
**Pull request by tautschnig**
Please review commit by commit - each of them solves a variable shadowing problem in some subset of classes/files.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/cpp/cpp_typecheck_virtual_table.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ void cpp_typecheckt::do_virtual_table(const symbolt &symbol)
       id2string(vt_symb_type.name) + "@"+ id2string(symbol.name);
     vt_symb_var.base_name=
       id2string(vt_symb_type.base_name) + "@" + id2string(symbol.base_name);
-    vt_symb_var.mode=ID_cpp;
```
Comment (smowton) [inline src/cpp/cpp_typecheck_virtual_table.cpp]: Unrelated change?

**Review by smowton — APPROVED** [review]

File: src/cpp/cpp_typecheck_virtual_table.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_virtual_table.cpp]: Hmm, yes, I'll remove it.

Comment (tautschnig) [reply src/cpp/cpp_typecheck_virtual_table.cpp]: Hmm, looking at my commit message that's actually explained:

    Do not shadow "mode" in cpp front-end

    Also pick up a previously set mode from a symbol rather than hard-coding ID_cpp
    where possible. The symbol's mode could be ID_C or ID_cpp, depending on the
    linkage specification.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cceb8f421ddc8c9854a961bb1b2e9a91c4719ff0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95450432

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a94836e6fa95b37ecf1d5e0626723e29f2171a84).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95509779

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e32dc1affe909276a1a7248b784ea763066194ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95526623

File: src/goto-cc/gcc_cmdline.h [hunk 2]
```diff
@@ -29,7 +29,7 @@ class gcc_cmdlinet:public goto_cc_cmdlinet
 protected:
   typedef std::vector<std::string> argst;
 
-  bool parse_arguments(const argst &args, bool in_spec_file);
+  bool parse_arguments(const argst &, bool in_spec_file);
```
Comment (owen-mc-diffblue) [inline src/goto-cc/gcc_cmdline.h]: Why leave the parameter name blank, instead of making it `args_to_parse` to match the definition?

**Review by owen-mc-diffblue — APPROVED** [review]

File: src/goto-cc/gcc_cmdline.h (hunk 2 above)
Comment (tautschnig) [reply src/goto-cc/gcc_cmdline.h]: Done as suggested!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2208e455f9e673004c31f0fb305805d161d5f9c2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95541818

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d077960b2d8299322c5dba4dae61f8cef3d1b34a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95543217
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: add43f119d9c25078d0f2933ebb259e4052fc601).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95548056

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bf4668bb114187a8240327ecef9646d2c8c45099).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95555454

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0c81f2749ac1daf3a49f476405f956019fb75a6e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95559606

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 519182af66f394b0feaff43206dddc15201f13a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95564288

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1e44ee67a7bbfc0dc855f5bff7c3d4702d06fa01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95565598

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 72c92977213a8308bb55350895e95af2cce6007f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95567949

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0d2c46535f81ef1a59e6ddabbc04b2ecc43e75e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95572787

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7e6a5531f2f56a6d5e9e8d64dc300cedfb3b5a6b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95587637

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c44b6009a5cb40b336f196860fd080aae541e898).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95588316

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bca4d587cdafb1e2ca058ba9674c7234ed833b5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95616873

File: src/cpp/cpp_typecheck_virtual_table.cpp (hunk 1 above)
Comment (chrisr-diffblue) [reply src/cpp/cpp_typecheck_virtual_table.cpp]: I'm a little uncomfortable with either the commit message, or the commit - the headline of the message 'Do not shadow "mode" in cpp front-end' sounds like a simple refactoring, yet the commit *also* has a functional change in there too (which admittedly is mentioned in the body of the message) - i.e. the handling of previously set modes. Might it be better to either split that change out in to a separate commit, or change the wording of this commit so that the headline is more along the lines of 'Improve handling of "mode" in cpp frontend' ?

**Review by chrisr-diffblue — APPROVED** [review]
One minor comment about the commit message on one of the commits, but otherwise happy.

File: src/cpp/cpp_typecheck_virtual_table.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_virtual_table.cpp]: I have taken the route of amending the commit message as suggested.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ff24c47e6e20231bc7dea05ebf7043bdaba2e1d1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96175388
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ccc895da0b7f9c22942b2c335f0a261f05d60a18).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96177102
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5b9a7189b2a151f03a78ebce98d7c2fc52b2f57a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96190522
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by peterschrammel — COMMENTED** [review]
There's a failing test: https://travis-ci.org/diffblue/cbmc/jobs/474933310#L5042 and lots of clang-format issues.

**Comment by tautschnig** [comment]
My apologies for letting this PR grow beyond reviewable size. I'm working on per-directory PRs and will eventually turn this one into one of them.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2da216c7d5e771a6779db788f2c9b52efe657e94).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96499375

File: src/solvers/refinement/string_refinement.cpp [hunk 3]
```diff
@@ -776,7 +776,7 @@ decision_proceduret::resultt string_refinementt::dec_solve()
   while((loop_bound_--)>0)
   {
     dependencies.clean_cache();
-    const decision_proceduret::resultt res=supert::dec_solve();
+    res = supert::dec_solve();
```
Comment (thk123) [inline src/solvers/refinement/string_refinement.cpp]: I'm not a big fan of this change - makes it hard to see that this is constant throughout the loop and not dependent on it's outside loop state. A nice fix would of course be to introduce some functions or something, but perhaps an easier immediate fix for the shadowing would be call this  `loop_res` and the other one `initial_res`

File: src/solvers/refinement/string_constraint_generator_comparison.cpp [hunk 4]
```diff
@@ -191,7 +191,7 @@ string_constraint_generatort::add_axioms_for_hash_code(
   array_poolt &pool)
 {
   PRECONDITION(f.arguments().size() == 1);
-  string_constraintst constraints;
+  string_constraintst hash_constraints;
```
Comment (thk123) [inline src/solvers/refinement/string_constraint_generator_comparison.cpp:194]: :pick: I'd probably rename the member variable to `global_constraints` or `all_constraints`

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 5]
```diff
@@ -422,10 +422,10 @@ std::pair<exprt, string_constraintst> add_axioms_from_float_scientific_notation(
     int_type);
   array_exprt conversion_factor_table(
     array_typet(float_type, conversion_factor_table_size));
-  for(const auto &f : two_power_e_over_ten_power_d_table_negatives)
-    conversion_factor_table.copy_to_operands(constant_float(f, float_spec));
-  for(const auto &f : two_power_e_over_ten_power_d_table)
-    conversion_factor_table.copy_to_operands(constant_float(f, float_spec));
+  for(const auto &neg : two_power_e_over_ten_power_d_table_negatives)
```
Comment (thk123) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: :pick: Don't use abbreviations `negative`, `position` - though obviously a massive improvement on what was there before. 
:pick: Rename the param `f` to `float_expr`?

**Review by thk123 — APPROVED** [review]

File: src/solvers/refinement/string_constraint_generator_comparison.cpp (hunk 4 above)
Comment (tautschnig) [reply src/solvers/refinement/string_constraint_generator_comparison.cpp:194]: Please forgive me for not intending to act on this one: from reading the code I got the impression that the two cases touched here affect methods that should likely be deprecated or removed. @romainbrenguier Any insights maybe?

Comment (thk123) [reply src/solvers/refinement/string_constraint_generator_comparison.cpp:194]: No problem - all of these comments are optional, I'm not overly familiar with the code being modified here anyway.

File: src/solvers/refinement/string_refinement.cpp (hunk 3 above)
Comment (tautschnig) [reply src/solvers/refinement/string_refinement.cpp]: Renaming done as suggested (using `initial_result`, `refined_result`).

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 5 above)
Comment (tautschnig) [reply src/solvers/refinement/string_constraint_generator_float.cpp]: Done, and a separate commit doing the renaming of `f` is being prepared.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6e271c373cb3f32c5061585de9b9802777b2defe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96602029
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: adde39e177138d3f181fbffa72e46feb666090ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96604752

File: src/solvers/refinement/string_constraint_generator_comparison.cpp (hunk 4 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator_comparison.cpp:194]: Yes hash code and intern should be implemented via models rather than internally in the solver (they are very language dependent), so this should disappear.

**Review by romainbrenguier — APPROVED** [review]
